timemachine/cpp/src/kernels/k_exchange.cu

```diff
@@ -59,76 +65,47 @@ void __global__ k_setup_proposals(
-    double *__restrict__ dest_coords,            // [N, 3]
-    size_t *__restrict__ num_accepted            // [1]
+void __global__ k_accepted_exchange_move(
+    const int batch_size,
+    const int num_atoms_in_each_mol,
+    const int *__restrict__ accepted_batched_move, // [1]
+    const int *__restrict__ mol_idx_per_batch,     // [batch_size]
+    const int *__restrict__ mol_offsets,           // [num_target_mols]
+    const double *__restrict__ moved_coords,       // [batch_size, num_atoms_in_each_mol, 3]
+    double *__restrict__ dest_coords,              // [N, 3]
+    size_t *__restrict__ num_accepted,             // [1],
+    int *__restrict__ rand_offset                  // [1]
 ) {
     // Note that this kernel does not handle multiple proposals, expects that the proposals
     // have been reduced down to a single proposal beforehand.
-    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    const int idx = blockIdx.x * blockDim.x + threadIdx.x;
     assert(idx == 0);
 
-    const int moved_mol_idx = selected_mol[0];
-    const int mol_start = mol_offsets[moved_mol_idx];
-    const int mol_end = mol_offsets[moved_mol_idx + 1];
-    const int num_atoms_in_each_mol = mol_end - mol_start;
-
-    const RealType before_log_prob =
-        convert_nan_to_inf<RealType>(compute_logsumexp_final<RealType>(before_max[0], before_log_sum[0]));
-    const RealType after_log_prob =
-        convert_nan_to_inf<RealType>(compute_logsumexp_final<RealType>(after_max[0], after_log_sum[0]));
-
-    const RealType log_acceptance_prob = min(before_log_prob - after_log_prob, static_cast<RealType>(0.0));
-    const bool accepted = rand[0] < exp(log_acceptance_prob);
-    if (idx == 0 && accepted) {
+    const int batch_idx = accepted_batched_move[0];
+    // If the selected batch idx is not less then the total batch size, no proposal was accepted, we can exit immediately.
+    if (batch_idx >= batch_size) {
+        rand_offset[0] += batch_size;
+        return;
+    }
+    const int mol_idx = mol_idx_per_batch[batch_idx];
+    const int mol_start = mol_offsets[mol_idx];
+    // Increment offset by the index + 1, IE if the zeroth item in the batch was accepted offset by 1
+    rand_offset[0] += batch_idx + 1;
+    if (threadIdx.x == 0) {
         num_accepted[0]++;
     }
 
-    if (accepted && idx == 0) {
-        // If accepted, move the coords of the selected mol into place
-        for (int i = 0; i < num_atoms_in_each_mol; i++) {
-            dest_coords[(mol_start + i) * 3 + 0] = moved_coords[i * 3 + 0];
-            dest_coords[(mol_start + i) * 3 + 1] = moved_coords[i * 3 + 1];
-            dest_coords[(mol_start + i) * 3 + 2] = moved_coords[i * 3 + 2];
-        }
+    // If accepted, move the coords of the selected mol into place
+    for (int i = 0; i < num_atoms_in_each_mol; i++) {
+        auto moved_x = moved_coords[num_atoms_in_each_mol * batch_idx * 3 + i * 3 + 0];
```

Comment:
auto here means double? (should be explicit)

Reply:
Woops from some debugging that wasn't undone, should just copy directly without the intermediate variable

Reply:
Removed in https://github.com/proteneer/timemachine/pull/1274/commits/21ccc9876c862db3f9704074b4f032834f9ec3fd